S4 t4 v2/implement db recurring schedule

prdbondi opened the pull request:
**Scrum Board Ticket**: S4 T4: Implement DB logic for adding recurring weekly schedule

**Story**: Story 4 Add same recurring weekly schedule

**Description**: Implementing an api endpoint to add worker timeslots into the database. Code submitted for PR review prior to functionality being reached due to blockage on stub branch: https://github.com/sehersarin/MSCI342-Term-Project/pull/6. 
Please review and inform of any errors / omissions. 
Note that due to time constraints the current version of the software only aims to add single timeslots, as opposed to recurring timeslots on a weekly basis. 

**Testing Criteria**: Once blocking PR is merged, can testcode by running locally and sending requests to the /api/worker-availability endpoint for valid and invalid query values. Can also ensure that all automated tests in Jest pass and have 100% file coverage.

**How Reviewers can Test Code:**

Run the code locally and send requests to the /api/worker-availability endpoint for valid and invalid query parameter values (e.g., invalid workerId, valid workerId).
Run the automated tests locally and ensure all tests pass and have 100% file coverage.

prdbondi commented on server/models/db/add-worker-timeslot.js:
```diff
@@ -0,0 +1,17 @@
+async function insertWorkerTimeslot(workerTimeslotId, slotId, schoolId, workerId, status, date) {
+    // Added conditional statements to prevent purpose, studentNotes, and workerComments from storing "null" instead of null.    
+    // Assumes all new appointments are upcoming and not for the past (as front end only allows users to book appointments for future dates).
+    return db.any(`insert into ${Tables.worker_timeslot} (worker_timeslot_id, slot_id, school_id, worker_id, status, date) values (${workerTimeslotId}, ${slotId}, ${schoolId}, ${workerId}, ${status ? `'${status}'` : null}, ${date ? `'${date}'` : null}, '${AddTimeslotStatus.available}');`);
```
Please note the ambiguity surrounding the date data type in this method. 
I am unsure exactly how to format the date parameter into the statement which inserts the record into the database

catai9 replied:
I would leverage moment to format the date so it is consistent in the database table. Refer to the documentation here: https://momentjs.com/

catai9 commented on server/constants/timeslot-status.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "available": "available",
+    "unavailable": "unavailable"}
```
Minor coding best practice: I would recommend moving the final curly braces to the next line to comply with typical json file structures.

prdbondi replied:
Thanks Amy!

catai9 commented on server/models/data/Timeslot.js:
```diff
@@ -0,0 +1,12 @@
+const UserTypes = require('../../constants/userTypes.json');
+
+class Timeslot {
+    // The password field will NOT be stored outside of the database for security purposes.
```
Not sure if this comment is relevant for the specific code lines. It may be better to remove it to avoid confusion.

prdbondi replied:
Will do, good point

catai9 commented on server/models/db/add-worker-timeslot.js:
```diff
@@ -0,0 +1,18 @@
+const { db } = require('../../lib/connection');
+
+const Tables = require('../../constants/tables.json');
+const AddTimeslotStatus = require('../../constants/timeslot-status.json');
```
I would rename this to TimeslotStatus to avoid confusion as it contains the possible values of the timeslot status rather than performs a manipulation.

prdbondi replied:
Appreciate the level of detail. Very helpful!

catai9 commented on server/models/db/add-worker-timeslot.js:
```diff
@@ -0,0 +1,18 @@
+const _ = require('lodash');
```
I would rename this file to be `schedule` or `timeslot` or another noun rather than `add-worker-timeslot` to increase flexibility in the future as other worker timeslot actions, such as viewing or fetching specific timeslots may be performed in the future and should be located within the same file for simplicity.

prdbondi replied:
Understood. Scability is important.

catai9 commented on server/models/handlers/workerTimeslot.js:
```diff
@@ -0,0 +1,16 @@
+const timeslotModel = require('../db/add-worker-timeslot');
+
+// Adds an appointment to the database and returns true upon successful completion.
```
I would revise this comment as it may be confusing since this method actually inserts worker availability rather than adds an appointment to the database.

prdbondi replied:
My mistake, thank you for the highlight.

catai9 reviewed (changes requested):
Hi Peter, 
You've made really great progress in your code. I have suggested a few minor changes and would also suggest that automated unit tests be added for your specific handler method. Please refer to the ones in the `test` folder for some examples.

prdbondi commented:
Amy, appreciate the great support and detailed comments, they are very helpful. 
I have rectified the smaller edits and have started working on the moment.js date formatting as well as the automated unit tests. You should see progress on these by tomorrow morning 10am at the latest.

prdbondi commented:
Utilized moment date formatting and added automated Test code. Encountering checksum error when trying to run. Hoping to resolve before 10pm.

prdbondi commented:
Reverted moment changes as they were causing issues. Ran npm test to best of knowledge and fixed any syntax errors. Functionality affected by system setup.

prdbondi commented:
Closing PR as functionality not proven by sprint deadline. Hope to resume work in Sprint 2
